src/Workspaces/Core/Portable/Utilities/WeakSet`1.cs

```diff
@@ -0,0 +1,37 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections.Generic;
+
+namespace Roslyn.Utilities
+{
+    /// <summary>
+    /// A simple collection of values held as weak references. Objects in the set are compared by reference equality.
+    /// </summary>
+    /// <typeparam name="T">The type of object stored in the set.</typeparam>
+    internal sealed class WeakSet<T>
+        where T : class?
+    {
+        private readonly HashSet<ReferenceHolder<T>> _values = new HashSet<ReferenceHolder<T>>();
+
+        public WeakSet()
+        {
+        }
+
+        public bool Add(T value)
+        {
+            if (Contains(value))
```

Comment:
is there a benefit to this check?  esp as the existing ref could release, and your system would say that it's still there, and the new value would not be added, right?

Reply:
It avoids the creation of a weak reference if a weak reference for this object is already being tracked.

> esp as the existing ref could release, and your system would say that it's still there, and the new value would not be added, right?

I don't understand the sequence here.

Reply:
> as the existing ref could release

This is not possible. The reference is being kept alive by `value` which is used by the call to `Weak` after this check.

Reply:
that's fair here.  teh 'value' coming in will never be nulled out.  so it should never match.